File: app/src/main/java/org/schabi/newpipe/player/mediaitem/PlaceholderTag.java
package org.schabi.newpipe.player.mediaitem;

import org.schabi.newpipe.extractor.stream.StreamType;
import org.schabi.newpipe.util.Constants;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * This is a Placeholding {@link MediaItemTag}, designed as a dummy metadata object for
 * any stream that has not been resolved.
 *
 * This object cannot be instantiated and does not hold real metadata of any form.
 * */
public final class PlaceholderTag implements MediaItemTag {
    public static final PlaceholderTag EMPTY = new PlaceholderTag(null);
    private static final String UNKNOWN_VALUE_INTERNAL = "Placeholder";

    @Nullable
    private final Object extras;

    private PlaceholderTag(@Nullable final Object extras) {
        this.extras = extras;
    }

    @NonNull
    @Override
    public List<Exception> getErrors() {
        return Collections.emptyList();
    }

    @Override
    public int getServiceId() {
        return Constants.NO_SERVICE_ID;
    }

    @Override
    public String getTitle() {
        return UNKNOWN_VALUE_INTERNAL;
    }

    @Override
    public String getUploaderName() {
        return UNKNOWN_VALUE_INTERNAL;
    }

    @Override
    public long getDurationSeconds() {
        return 0;
    }

    @Override
    public String getStreamUrl() {
        return UNKNOWN_VALUE_INTERNAL;
    }

    @Override
    public String getThumbnailUrl() {
        return UNKNOWN_VALUE_INTERNAL;
    }

    @Override
    public String getUploaderUrl() {
        return UNKNOWN_VALUE_INTERNAL;
    }

    @Override
    public StreamType getStreamType() {
        return StreamType.NONE;
    }

    @Override
    public <T> Optional<T> getMaybeExtras(@NonNull final Class<T> type) {
        return Optional.ofNullable(extras).map(type::cast);
    }

    @Override
    public <T> MediaItemTag withExtras(@NonNull final T extra) {
        return new PlaceholderTag(extra);
    }
}

File: app/src/main/java/org/schabi/newpipe/player/mediaitem/PlaceholderTag.kt
package org.schabi.newpipe.player.mediaitem

import org.schabi.newpipe.extractor.stream.StreamType
import org.schabi.newpipe.util.NO_SERVICE_ID

/**
 * This is a Placeholding [MediaItemTag], designed as a dummy metadata object for
 * any stream that has not been resolved.
 *
 * This object cannot be instantiated and does not hold real metadata of any form.
 */
class PlaceholderTag private constructor(private val extras: Any? = null) : MediaItemTag {

    override val errors: List<Exception> = emptyList()

    override val serviceId: Int = NO_SERVICE_ID

    override val title: String = UNKNOWN_VALUE_INTERNAL

    override val uploaderName: String = UNKNOWN_VALUE_INTERNAL

    override val durationSeconds: Long = 0

    override val streamUrl: String = UNKNOWN_VALUE_INTERNAL

    override val thumbnailUrl: String = UNKNOWN_VALUE_INTERNAL

    override val uploaderUrl: String = UNKNOWN_VALUE_INTERNAL

    override val streamType: StreamType = StreamType.NONE

    override fun <T : Any> getMaybeExtras(type: Class<T>): T? {
        return type.cast(extras)
    }

    override fun <T : Any> withExtras(extra: T): MediaItemTag {
        return PlaceholderTag(extra)
    }

    companion object {
        @JvmField
        val EMPTY = PlaceholderTag()

        private const val UNKNOWN_VALUE_INTERNAL = "Placeholder"
    }
}
